File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/camera2/interop/Camera2CameraControlImpl.kt
package dev.hebei.camerax_android.camera2.interop

import androidx.camera.camera2.interop.Camera2CameraControl
import androidx.camera.camera2.interop.CaptureRequestOptions
import androidx.camera.core.CameraControl
import androidx.core.content.ContextCompat
import com.google.common.util.concurrent.FutureCallback
import com.google.common.util.concurrent.Futures
import dev.hebei.camerax_android.CameraXRegistrar
import dev.hebei.camerax_android.PigeonApiCamera2CameraControlApi

class Camera2CameraControlImpl(private val registrar: CameraXRegistrar) : PigeonApiCamera2CameraControlApi(registrar) {
    override fun from(cameraControl: CameraControl): Camera2CameraControl {
        return Camera2CameraControl.from(cameraControl)
    }

    override fun addCaptureRequestOptions(
        pigeon_instance: Camera2CameraControl, bundle: CaptureRequestOptions, callback: (Result<Unit>) -> Unit
    ) {
        val future = pigeon_instance.addCaptureRequestOptions(bundle)
        val executor = ContextCompat.getMainExecutor(registrar.context)
        Futures.addCallback(future, object : FutureCallback<Void> {
            override fun onSuccess(result: Void?) {
                callback(Result.success(Unit))
            }

            override fun onFailure(t: Throwable) {
                callback(Result.failure(t))
            }
        }, executor)
    }

    override fun clearCaptureRequestOptions(pigeon_instance: Camera2CameraControl, callback: (Result<Unit>) -> Unit) {
        val future = pigeon_instance.clearCaptureRequestOptions()
        val executor = ContextCompat.getMainExecutor(registrar.context)
        Futures.addCallback(future, object : FutureCallback<Void> {
            override fun onSuccess(result: Void?) {
                callback(Result.success(Unit))
            }

            override fun onFailure(t: Throwable) {
                callback(Result.failure(t))
            }
        }, executor)
    }

    override fun getCaptureRequestOptions(pigeon_instance: Camera2CameraControl): CaptureRequestOptions {
        return pigeon_instance.captureRequestOptions
    }

    override fun setCaptureRequestOptions(
        pigeon_instance: Camera2CameraControl, bundle: CaptureRequestOptions, callback: (Result<Unit>) -> Unit
    ) {
        val future = pigeon_instance.setCaptureRequestOptions(bundle)
        val executor = ContextCompat.getMainExecutor(registrar.context)
        Futures.addCallback(future, object : FutureCallback<Void> {
            override fun onSuccess(result: Void?) {
                callback(Result.success(Unit))
            }

            override fun onFailure(t: Throwable) {
                callback(Result.failure(t))
            }
        }, executor)
    }
}

File: camerax_android/lib/src/camera2/interop/camera2_camera_control_impl.dart
import 'package:camerax_android/src/camera2.dart';
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class Camera2CameraControlImpl implements Camera2CameraControl {
  final Camera2CameraControlProxyApi api;

  Camera2CameraControlImpl.internal(this.api);

  @override
  Future<void> addCaptureRequestOptions(CaptureRequestOptions bundle) =>
      api.addCaptureRequestOptions(bundle.api);

  @override
  Future<void> clearCaptureRequestOptions() => api.clearCaptureRequestOptions();

  @override
  Future<CaptureRequestOptions> getCaptureRequestOptions() =>
      api.getCaptureRequestOptions().then((e) => e.impl);

  @override
  Future<void> setCaptureRequestOptions(CaptureRequestOptions bundle) =>
      api.setCaptureRequestOptions(bundle.api);
}

final class Camera2CameraControlChannelImpl
    extends Camera2CameraControlChannel {
  @override
  Camera2CameraControl from(CameraControl cameraControl) {
    final api = Camera2CameraControlProxyApi.from(
      cameraControl: cameraControl.api,
    );
    return Camera2CameraControlImpl.internal(api);
  }
}
